app/Main.hs

```diff
@@ -1,150 +1,40 @@
 module Main where
 
-import Control.Monad.State
-import System.IO (hSetBuffering, BufferMode(NoBuffering), stdout)
-import System.Console.ANSI
-import System.ProgressBar
-import Text.Printf
-import Unused.TermSearch (search)
-import Unused.Parser (parseLines, ParseError)
-import Unused.Types
+import Options.Applicative
+import Unused.Parser (parseLines)
+import Unused.Types (withOneOccurrence, withOneFile)
+import Unused.CLI (SearchRunner(..), executeSearch, printParseError, printSearchResults, resetScreen)
+
+data Options = Options
+    { oSearchRunner :: SearchRunner
+    }
 
 main :: IO ()
-main = do
+main = run =<< execParser
+    (parseOptions `withInfo` "Analyze potentially unused code")
+
+run :: Options -> IO ()
+run options = do
     terms <- pure . lines =<< getContents
-    results <- executeSearch terms
+    results <- executeSearch (oSearchRunner options) terms
     let response = parseLines $ unlines results
 
     resetScreen
-    case withOneOccurrence $ withOneFile response of
-        Right termMatchSet -> do
-            mapM_ (printDirectorySection $ maxWidth) responses
-          where
-            responses = responsesGroupedByPath termMatchSet
-            allSets = listFromMatchSet =<< map snd responses
-            allResults = fmap snd allSets
-            termLength = return . length . tmTerm
-            maxWidth = maximum $ termLength =<< trMatches =<< allResults
-        Left e ->
-            printParseError e
-
-    return ()
-
-executeSearch :: [String] -> IO [String]
-executeSearch terms = do
-    resetScreen
-    printAnalysisHeader terms
-    (results, _) <- runStateT (performSearch $ length terms) terms
-    resetScreen
-
-    return $ concat results
-
-performSearch :: Int -> StateT [String] IO [[String]]
-performSearch total = do
-    currentTerm <- getSearchTerm
-    searchResults <- liftIO $ search currentTerm
-
-    remainingTerms <- get
-    let remaining = length remainingTerms
-
-    liftIO $ printProgressBar (total - remaining) total
-
-    if remaining > 0
-        then do
-            res <- performSearch total
-            return $ searchResults:res
-        else return [searchResults]
-
-getSearchTerm :: StateT [String] IO String
-getSearchTerm = do
-    (x:xs) <- get
-    put xs
-    return x
-
-printProgressBar :: Int -> Int -> IO ()
-printProgressBar complete total = do
-    let message = "Working"
-    let progressBarWidth = 60
 
-    hSetBuffering stdout NoBuffering
-    progressBar (msg message) percentage progressBarWidth (toInteger complete) (toInteger total)
+    either printParseError printSearchResults $
+        (withOneOccurrence $ withOneFile response)
```

Comment:
I believe these parens are are redundant.

Shameless plug: you should set up Code Climate with the hlint engine (or just use hlint locally) :)